infrastructure/keymanager/src/keymanager.rs

```diff
@@ -0,0 +1,262 @@
+        KeyManager { master_key: SecretKey::random(&mut rng), branch_seed: "".to_string(), primary_key_index: 0 }
+    }
+
+    /// Constructs a KeyManager from known parts
+    pub fn from(master_key: SecretKey, branch_seed: String, primary_key_index: usize) -> KeyManager {
+        KeyManager { master_key, branch_seed, primary_key_index }
+    }
+
+    /// Constructs a KeyManager by generating a master_key=SHA256(seed_phrase) using a non-mnemonic seed phrase
+    pub fn from_seed_phrase(
+        seed_phrase: String,
+        branch_seed: String,
+        primary_key_index: usize,
+    ) -> Result<KeyManager, KeyManagerError>
+    {
+        match SecretKey::from_bytes(sha256(seed_phrase.into_bytes()).as_slice()) {
+            Ok(master_key) => Ok(KeyManager { master_key, branch_seed, primary_key_index }),
+            Err(e) => Err(KeyManagerError::from(e)),
+        }
+    }
+
+    /// Creates a KeyManager from the provided sequence of mnemonic words, the language of the mnemonic sequence will be
+    /// auto detected
+    pub fn from_mnemonic(
+        mnemonic_seq: &Vec<String>,
+        branch_seed: String,
+        primary_key_index: usize,
+    ) -> Result<KeyManager, KeyManagerError>
+    {
+        match SecretKey::from_mnemonic(mnemonic_seq) {
+            Ok(master_key) => Ok(KeyManager { master_key, branch_seed, primary_key_index }),
+            Err(e) => Err(KeyManagerError::from(e)),
+        }
+    }
+
+    // TODO: file should be decrypted using Salsa20 or ChaCha20
+    /// Load KeyManager state from backup file
+    pub fn from_file(filename: &String) -> Result<KeyManager, KeyManagerError> {
+        let mut file_handle = match File::open(&filename) {
+            Ok(file) => file,
+            Err(_e) => return Err(KeyManagerError::FileOpen),
+        };
+        let mut file_content = String::new();
+        match file_handle.read_to_string(&mut file_content) {
+            Ok(_) => match serde_json::from_str(&file_content) {
+                Ok(km) => Ok(km),
+                Err(_) => Err(KeyManagerError::Deserialize),
+            },
+            Err(_) => Err(KeyManagerError::FileRead),
+        }
+    }
+
+    /// Derive a new private key from master key: derived_key=SHA256(master_key||branch_seed||index)
+    pub fn derive_key(&self, key_index: usize) -> Result<DerivedKey, ByteArrayError> {
+        let concatenated = format!("{}{}", self.master_key.to_hex(), key_index.to_string());
+        match SecretKey::from_bytes(sha256(concatenated.into_bytes()).as_slice()) {
+            Ok(k) => Ok(DerivedKey { k, key_index }),
+            Err(e) => Err(e),
+        }
+    }
+
+    /// Generate next deterministic private key derived from master key
+    pub fn next_key(&mut self) -> Result<DerivedKey, ByteArrayError> {
+        self.primary_key_index += 1;
+        (self.derive_key(self.primary_key_index))
+    }
+
+    // TODO: file should be encrypted using Salsa20 or ChaCha20
```

Comment:
Same as above. Rather think about keeping coupling as loose as possible

Reply:
Decoupled file backup from KeyManager, should be able to save and load any struct that implements serde serialisation and deserialisation.